service/worker/scanner/replication/replication.go

```diff
@@ -0,0 +1,263 @@
+	}
+
+	shardCount := metadataResp.ShardCount
+	skipAfter := options.SkipAfterTime
+	if skipAfter.IsZero() {
+		skipAfter = workflow.Now(ctx2)
+	}
+	var futures []workflow.Future
+	batchSize := (shardCount + concurrentCount - 1) / concurrentCount
+	for beginShardID := int32(1); beginShardID <= shardCount; beginShardID += batchSize {
+		endShardID := beginShardID + batchSize - 1
+		if endShardID > shardCount {
+			endShardID = shardCount
+		}
+		rangeRequest := genReplicationForShardRange{
+			BeginShardID:  beginShardID,
+			EndShardID:    endShardID,
+			NamespaceID:   metadataResp.NamespaceID,
+			SkipAfterTime: skipAfter,
+		}
+		future := workflow.ExecuteActivity(ctx2, a.GenerateReplicationTasks, rangeRequest)
+		futures = append(futures, future)
+	}
+
+	for _, f := range futures {
+		if err := f.Get(ctx2, nil); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// GetMetadata returns history shard count and namespaceID for requested namespace.
+func (a *activities) GetMetadata(ctx context.Context, request metadataRequest) (*metadataResponse, error) {
+	nsEntry, err := a.namespaceRegistry.GetNamespace(namespace.Name(request.Namespace))
+	if err != nil {
+		return nil, err
+	}
+
+	return &metadataResponse{
+		ShardCount:  a.historyShardCount,
+		NamespaceID: string(nsEntry.ID()),
+	}, nil
+}
+
+// GenerateReplicationTasks generates replication task for last history event for each workflow.
+func (a *activities) GenerateReplicationTasks(ctx context.Context, request genReplicationForShardRange) error {
+	perShard := genReplicationForShard{
+		ShardID:       request.BeginShardID,
+		NamespaceID:   request.NamespaceID,
+		SkipAfterTime: request.SkipAfterTime,
+	}
+	var progress heartbeatProgress
+	if activity.HasHeartbeatDetails(ctx) {
+		if err := activity.GetHeartbeatDetails(ctx, &progress); err == nil {
+			perShard.ShardID = progress.ShardID
+			perShard.PageToken = progress.PageToken
+			perShard.Index = progress.Index
+		}
+	}
+	for ; perShard.ShardID <= request.EndShardID; perShard.ShardID++ {
+		if err := a.genReplicationTasks(ctx, perShard); err != nil {
+			return err
+		}
+		// heartbeat progress only apply for first shard
+		perShard.PageToken = nil
+		perShard.Index = 0
+	}
+	return nil
+}
+
+func (a *activities) genReplicationTasks(ctx context.Context, request genReplicationForShard) error {
+	pageToken := request.PageToken
+	startIndex := request.Index
+	for {
+		var listResult *persistence.ListConcreteExecutionsResponse
+		op := func(ctx context.Context) error {
+			var err error
+			listResult, err = a.executionManager.ListConcreteExecutions(&persistence.ListConcreteExecutionsRequest{
+				ShardID:   request.ShardID,
+				PageSize:  listExecutionPageSize,
+				PageToken: pageToken,
+			})
+			return err
+		}
+
+		err := backoff.RetryContext(ctx, op, common.CreateHistoryServiceRetryPolicy(), common.IsPersistenceTransientError)
+		if err != nil {
+			return err
+		}
+
+		for i := startIndex; i < len(listResult.States); i++ {
```

Comment:
plz consider using a iterator
giant for loops are hard to maintain
ref: https://github.com/temporalio/temporal/blob/744f26d/service/history/nDCStateRebuilder.go#L116-L132